percona/controllers/service/service.go

```diff
@@ -0,0 +1,230 @@
+package service
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/percona/percona-postgresql-operator/internal/kubeapi"
+	crv1 "github.com/percona/percona-postgresql-operator/pkg/apis/crunchydata.com/v1"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+type ServiceType string
+
+const (
+	PGPrimaryServiceType = ServiceType("primary")
+	PGBouncerServiceType = ServiceType("bouncer")
+	PGReplicaServiceType = ServiceType("replica")
+)
+
+func CreateOrUpdate(clientset kubeapi.Interface, cluster *crv1.PerconaPGCluster, svcType ServiceType) error {
+	ctx := context.TODO()
+	var service *corev1.Service
+	switch svcType {
+	case PGPrimaryServiceType:
+		svc, err := getPrimaryServiceObject(cluster)
+		if err != nil {
+			return errors.Wrap(err, "get primary service object")
+		}
+		service = svc
+	case PGBouncerServiceType:
+		svc, err := getPGBouncerServiceObject(cluster)
+		if err != nil {
+			return errors.Wrap(err, "get pgBouncer service object")
+		}
+		service = svc
+	case PGReplicaServiceType:
+		svc, err := getReplicaServiceObject(cluster)
+		if err != nil {
+			return errors.Wrap(err, "get replica service object")
+		}
+		service = svc
+	}
+	oldSvc, err := clientset.CoreV1().Services(cluster.Namespace).Get(ctx, service.Name, metav1.GetOptions{})
+	if err != nil {
+		_, err = clientset.CoreV1().Services(cluster.Namespace).Create(ctx, service, metav1.CreateOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "create service %s", service.Name)
+		}
+		return nil
+	}
+	service.Spec.ClusterIP = oldSvc.Spec.ClusterIP
+	if reflect.DeepEqual(service.Spec, oldSvc.Spec) {
+		return nil
+	}
+	service.ResourceVersion = oldSvc.ResourceVersion
```

Comment:
resource version is being checked too, please move it before comparing svc specs

Reply:
`ResourceVersion` is a part of `ObectMeta`, but we comparing `Spec`